### PR TITLE
chore: improve SQL parsing

### PR DESCRIPTION
**Pull request by betodealmeida**
### SUMMARY

This is the first PR in a series to improve our SQL parsing, in effort to clean up the code base and make it more secure.

Currently, we have a class called `superset.sql_parse.ParsedQuery` that is used as an interface every time we need to parse or manipulate SQL. There are a few problems with the status quo:

1. Even though the class is called `ParsedQuery`, it's designed to accept a single SQL **statement**, not necessarily a query. For example, it has a `set_or_update_query_limit` method that works only on the first statement, if there are multiple.
2. Even though it's designed to accept a single statement, it doesn't raise an error when instantiated with a query that has multiple statements, so in several places the class is used with a full query.
3. There are multiple places that import `sqlparse` directly to manipulate SQL. For example, when inserting RLS into a query, or when modifying a query to insert a `LIMIT` or `TOP`.
4. In many places we manipulate SQL inefficiently, serializing and deserializing the SQL multiple times.

Ideally we'd have a single abstraction that handles SQL parsing. Superset code should never parse code directly by importing a 3rd party library like `sqlparse`, `sqloxide`, or `sqlglot`; instead, it should only call the abstraction. Similarly, the abstraction should wrap library specific exceptions (like `sqlglot.errors.ParseError`) with its own.

This PR introduces two new classes that aim to be those abstractions: `SQLScript` and `SQLStatement`. The `SQLStatement` strictly represents a single SQL statement. These classes are instantiated with SQL and an optional engine (a `BaseEngineSpec.engine` attribute), and should provide all the methods needed for manipulating and introspecting SQL queries and statements. The initial implementation uses `sqlglot`, but the clean interface allows the parser to be changed in the future if we want.

My plan for improving the SQL parsing is first to get rid of code outside of `sql_parse.py` that using `sqlparse` directly:

- [x] Introduce `SQLScript` and `SQLStatement`, and replace any simple calls to `sqlparse` with calls to the new classes (this PR).
- [ ] Implement a custom `SQLStatement` class for `Kusto`, since [it doesn't use conventional SQL](https://learn.microsoft.com/en-us/azure/data-explorer/kusto/query/).
- [ ] Implement `insert_rls_in_predicate` inside `SQLStatement`.
- [ ] Implement `get_cte_query` inside `SQLStatement`.
- [ ] Implement `apply_top_to_sql` in `SQLStatement`.

At this point we should have `sqlparse` being used only inside `sql_parse.py` by `ParsedQuery`. The last steps would be:

- [ ] Replace all calls to `ParsedQuery` with calls to the new classes.
- [ ] Get rid of the `sqlparse` dependency.

Note that despite the big changes, no public interfaces will change. Also, we've been gradually changing the parser due to security issues, having introduced `sqlglot` as a dependency in https://github.com/apache/superset/pull/26476. Because of this, I thought it wouldn't be necessary to write a SIP. I'm happy to write one if people think it's necessary.

### BEFORE/AFTER SCREENSHOTS OR ANIMATED GIF

N/A

### TESTING INSTRUCTIONS

Updated unit tests, with mostly cosmetic changes because the SQL pretty-printing has changed.

### ADDITIONAL INFORMATION

- [ ] Has associated issue:
- [ ] Required feature flags:
- [ ] Changes UI
- [ ] Includes DB Migration (follow approval process in [SIP-59](https://github.com/apache/superset/issues/13351))
  - [ ] Migration is atomic, supports rollback & is backwards-compatible
  - [ ] Confirm DB migration upgrade and downgrade tested
  - [ ] Runtime estimates and downtime expectations provided
- [ ] Introduces new feature or API
- [ ] Removes existing feature or API

### CODE REVIEW

File: superset-frontend/src/SqlLab/actions/sqlLab.js
Comment (betodealmeida) [inline superset-frontend/src/SqlLab/actions/sqlLab.js:915]: This is a nice benefit of this change: we can now have DB-specific formatting.

File: superset/connectors/sqla/models.py
Comment (betodealmeida) [inline superset/connectors/sqla/models.py:1417]: This method is **identical** to the base class.

File: superset/connectors/sqla/models.py
Comment (betodealmeida) [inline superset/connectors/sqla/models.py:1478]: This method is **identical** to the base class.

File: superset/db_engine_specs/postgres.py
Comment (betodealmeida) [inline superset/db_engine_specs/postgres.py:286]: Look, ma! No regexes! 😄

File: superset/sql_parse.py
Comment (betodealmeida) [inline superset/sql_parse.py:256]: This function was originally a method in `ParsedQuery`.

File: superset/sql_parse.py
Comment (betodealmeida) [inline superset/sql_parse.py:304]: This function was originally a method in `ParsedQuery`.

**Review by mistercrunch — COMMENTED** [review]
Few comments:
- Let's remove `sqlparse` from setup.py and requirements. I noticed `pip-compile-multi` was SUPER slow last time I tried it
- about the description and now improved support for multi-statements, are all the cases covered by new tests in `sql_parse_tests.py`?

**Comment by villebro** [comment]
I think this is a welcome change and a move in the right direction, but I feel a big change like this should probably be discussed in the form of a SIP first. I anticipate the final implementation to end up looking similar to this PR, but it may be beneficial to discuss the change first before jumping directly into the code.

**Comment by betodealmeida** [comment]
> I think this is a welcome change and a move in the right direction, but I feel a big change like this should probably be discussed in the form of a SIP first. I anticipate the final implementation to end up looking similar to this PR, but it may be beneficial to discuss the change first before jumping directly into the code.

SIP here: https://github.com/apache/superset/issues/26786

**Comment by villebro** [comment]
Thanks @betodealmeida !

**Comment by betodealmeida** [comment]
https://github.com/apache/superset/issues/26786 approved!

File: superset/sql_parse.py
Comment (mistercrunch) [inline superset/sql_parse.py:366]: Looking at the class design in this module, this looks like it really extract the first statement from a set of statement. It may not be for this PR, but ideally the classes become a better representation of the real word. Maybe there's a notion of a `SqlScript` or `SqlStatementCollection`, that's an array of other SQL objects like `SqlQuery`, `SqlDMLCommand`, ...

Maybe there's benefit in having inheritance defined there where a SqlQuery and DmlCommand are both derivative or `SqlStatement`

File: tests/integration_tests/core_tests.py
Comment (mistercrunch) [inline tests/integration_tests/core_tests.py:540]: [suggesting] it could be good to have some sort of reusable `compare_sql(strict=False, case_sensitive=False, disregard_schema_prefix=True)` function that could be reused in unit tests. Maybe it's a method of `ParseQuery` (`is_identical` or `is_similar`)

**Review by mistercrunch — COMMENTED** [review]
Left some comments. At a high level the main thing for me would be to clarify the object model and inheritance scheme for the SQL-related utilities. At a high level thinking about this, it sits close to db_engine_spec or at least relate to it in some ways.

File: superset/sql_parse.py
Comment (betodealmeida) [reply superset/sql_parse.py:366]: Not just the first. The hierarchy is that a `SQLQuery` has 0+ `SQLStatement`s. When you pass a SQL query to `SQLQuery` it will split the query into statements, and instantiate one `SQLStatement` for each one. The `SQLStatement._parse_statement` validates that the SQL passed corresponds to exactly one statement, which is why it looks like it's extracting the first one.

This is done because today in the codebase there's an implicit assumption that a `ParsedQuery` represents a single statement, but because this is not enforced there are many places in the code that pass a full query to it when using methods that really don't care.

File: superset/sql_parse.py
Comment (betodealmeida) [reply superset/sql_parse.py:25]: Ah, I like this!

@john-bodley, you did some work on reorganizing the codebase, what are your thought here?

File: superset/sql_parse.py
Comment (betodealmeida) [reply superset/sql_parse.py:416]: This method is meant for statements that set parameters using the `SET foo = 'bar'` syntax, which seems to be more universal than `USE ...`. We have a few DB engine specs that need to know if `search_path` was set to something different for security reasons, which is why I added this method. And while today it's used just to check for the `search_path,` I used a more generic name because in the future it could be leveraged in different ways.

I agree that there's functionality here that is DB-specific, but I'm hoping to address that by always requiring the dialect to be specified together with the SQL when instantiating the `SQLStatement`, so we can leave the heavy work to `sqlglot`. If we need custom functionality we can always subclass the `SQLStatement` for specific dialects, though hopefully we'd be able to avoid that by implementing the dialect upstream in `sqlglot` directly.

The main reason for having DB-specific handling here and not in the DB engine specs is that I'd like for all the SQL parsing to be done only by these new classes, so we can have a cleaner interface. So I actually want to remove the custom DB engine spec methods that are related to SQL parsing that exist today, since the DB engine specs already do too much.

File: tests/integration_tests/core_tests.py
Comment (betodealmeida) [reply tests/integration_tests/core_tests.py:540]: Yeah, good point, I'll add that!

File: superset/sql_parse.py
Comment (eschutho) [reply superset/sql_parse.py:366]: I had a similar question, reading `SQLQuery` as solely a data retrieval query or a type of statement, rather than a collection of statements. `SQLInstruction` as another option?

File: superset/sql_parse.py
Comment (mistercrunch) [inline superset/sql_parse.py:404]: Super minor, but here you could fail faster (assuming that sqlglot.parse is a generator), meaning you could:
```python
has_statement = False
for statement in sqlglot.parse(sql_statement, dialect=dialect):
    if has_statement and statement:
        raise{...}
    else:
        has_statement = True
 ```

File: superset/sql_parse.py
Comment (mistercrunch) [reply superset/sql_parse.py:416]: Oh sorry I got confused on my previous comment, I read the `SET` as a `USE` and commented based on that. I guess they both have something in comment as they are context-related statements.

File: superset/sql_parse.py
Comment (betodealmeida) [reply superset/sql_parse.py:404]: Ah, good idea. Unfortunately I tested and it doesn't seem to be the case:

```python
import time

import sqlglot

padding = "a" * 100000000
start = time.time()
for stmt in sqlglot.parse(f"SELECT 'small'; SELECT 'la{padding}rge';"):
    print(time.time() - start)
```

```bash
$ python test-sqlglot.py
22.041950941085815
22.04202389717102
$
```

**Comment by rafehi** [comment]
Where are we at with this work? There's some additional changes that I'd be willing to take on (https://github.com/apache/superset/pull/19572) but looks like this is a dependency. 

We've made and validated significant performance improvements in our Superset fork, but makes sense to consolidate those into SQLScript and SQLStatement rather than directly replace sqlparse calls with sqlglot.

**Review by mistercrunch — APPROVED** [review]
LGTM

**Review by eschutho — APPROVED** [review]
I reviewed just the cypress test portion as a code-owner requirement. That LGTM. 👍

**Comment by mistercrunch** [comment]
Hey @betodealmeida , this PR might have broken a Hive-specific test -> https://github.com/apache/superset/actions/runs/8272398041/job/22634142747#step:11:608

It appears to fail after this PR was merged (I noticed since it was failing on some other unrelated PR) and trace it back here. Note that this particular GHA test for Hive runs only when `superset/**` is altered by a PR, that explains why other builds succeeded on `master` since then.

Hive seems to be complaining about a `GROUP BY 1` but from memory it totally supports that. Unclear what's up with the test but I thought you might know more given this PR.

**Pull request merged**
